src/GovUk.Education.ExploreEducationStatistics.Notifier/Functions/ReleaseNotifier.cs

```diff
@@ -0,0 +1,184 @@
+                var supersededPublicationTitle = msg.SupersededPublicationTitles[i];
+
+                var query = new TableQuery<SubscriptionEntity>()
+                    .Where(TableQuery.GenerateFilterCondition("PartitionKey", QueryComparisons.Equal,
+                        supersededPublicationId.ToString()));
+
+                logger.LogInformation("Emailing subscribers from a superseded publication");
+                var numSupersedeSubscriberEmailsSent = 0;
+
+                TableContinuationToken? token = null;
+                do
+                {
+                    var resultSegment =
+                        await table.ExecuteQuerySegmentedAsync(query, token);
+                    token = resultSegment.ContinuationToken;
+
+                    numSupersedeSubscriberEmailsSent += resultSegment.Results.Count;
+                    foreach (var entity in resultSegment.Results)
+                    {
+                        var email = entity.RowKey;
+
+                        if (sentToEmailAddresses.Contains(email))
+                        {
+                            numSupersedeSubscriberEmailsSent--;
+                            continue;
+                        }
+                        sentToEmailAddresses.Add(email);
+
+                        var unsubscribeToken =
+                            _tokenService.GenerateToken(tokenSecretKey, email, DateTime.UtcNow.AddYears(1));
+                        var values = new Dictionary<string, dynamic>
+                        {
+                            // Use values from the queue just in case the name or slug of the publication changes
+                            { "publication_name", msg.PublicationName },
+                            { "release_name", msg.ReleaseName },
+                            {
+                                "release_link",
+                                $"{webApplicationBaseUrl}find-statistics/{msg.PublicationSlug}/{msg.ReleaseSlug}"
+                            },
+                            {
+                                // NOTE: update_note not used by original release email template
+                                "update_note", msg.UpdateNote
+                            },
+                            { "unsubscribe_link", $"{baseUrl}{supersededPublicationId}/unsubscribe/{unsubscribeToken}" },
+                            { "superseded_publication_title", supersededPublicationTitle },
+                        };
+
+                        _emailService.SendEmail(notificationClient, email, newReleaseSupersededSubscribersTemplateId, values);
+                    }
+                } while (token != null);
+
+                logger.LogInformation("Emailed {NumEmailsSent} subscribers from a superseded publication",
+                    numSupersedeSubscriberEmailsSent);
+            }
+
+            logger.LogInformation("Sent {NumEmailsSent} emails in total to subscribers",
+                sentToEmailAddresses.Count);
+        }
```

Comment:
I'm not keen on this function because I think it has too many responsibilities now. It's got quite long to read.

I've not looked at it too closely but feels like there's a common method to be extracted between the regular and superseded functionality.

I think when we discussed the approach for this I mentioned my plan would be to have separate functions for the different cases, so there would be a function for sending the publication emails, and a function for sending the superseded emails. Each function would know about its own template id and have the responsibility for querying subscribers for the publication given in the message and building the appropriate email template values for that template. I think there would still be some common code reused between them for running the query.

I think it makes it a bit cleaner with it being clear which template each function is using and we know why each function has been invoked. It should be easier to unit test each function rather than one big one. If we also had variants of the functions for amendments we'd get rid of the problem where we set the `update_note` variable when it's not always relevant.

I don't have a preference on doing this though and it could be argued that it puts too much responsibility on the publisher to queue the correct type and number of messages but I don't see that as a big overhead.

Reply:
Changes have been made - most notable created methods `GetSubscriberEmails`/ `SendSubsciberEmail`/`SendSupersededSubscriberEmail`